crates/hulk_replayer/src/window.rs

```diff
@@ -14,56 +19,65 @@ use crate::{
 };
 
 pub struct Window {
-    replayer: Replayer<ReplayerHardwareInterface>,
+    replayer: Arc<Mutex<Replayer<ReplayerHardwareInterface>>>,
+    time_sender: watch::Sender<SystemTime>,
+    frames_changed: watch::Receiver<()>,
     frame_range: FrameRange,
     viewport_range: ViewportRange,
     position: RelativeTime,
 }
 
 impl Window {
-    pub fn new(replayer: Replayer<ReplayerHardwareInterface>) -> Self {
-        let frame_range = join_timing(&replayer);
+    pub fn new(
+        creation_context: &CreationContext,
+        replayer: Replayer<ReplayerHardwareInterface>,
+    ) -> Self {
+        let replayer = Arc::new(Mutex::new(replayer));
+        let (time_sender, time_receiver) = watch::channel(SystemTime::UNIX_EPOCH);
+        let (frames_changed_sender, frames_changed_receiver) = watch::channel(());
+        spawn_replay_thread(
+            replayer.clone(),
+            creation_context.egui_ctx.clone(),
+            time_receiver,
+            frames_changed_sender,
+        );
+
+        let frame_range = join_timing(&replayer.lock().unwrap());
         let viewport_range = ViewportRange::from_frame_range(&frame_range);
+
         Self {
             replayer,
+            time_sender,
+            frames_changed: frames_changed_receiver,
             frame_range,
             viewport_range,
             position: RelativeTime::new(0.0),
         }
     }
 
     fn replay_at_position(&mut self) {
-        let timestamp = self.position.map_to_absolute_time(&self.frame_range);
-        let recording_indices = self.replayer.get_recording_indices_mut();
-        let frames = recording_indices
-            .into_iter()
-            .map(|(name, index)| {
-                (
-                    name,
-                    index
-                        .find_latest_frame_up_to(timestamp.inner())
-                        .expect("failed to find latest frame"),
-                )
-            })
-            .collect::<BTreeMap<_, _>>();
-        for (name, frame) in frames {
-            if let Some(frame) = frame {
-                self.replayer
-                    .replay(&name, frame.timing.timestamp, &frame.data)
-                    .expect("failed to replay frame");
-            }
-        }
+        let frame_range = join_timing(&self.replayer.lock().unwrap());
+        let timestamp = self.position.map_to_absolute_time(&frame_range);
```

Comment:
Hmmm, maybe we need to rethink for a second which party is responsible for holding which state and when we update what. What I would feel as a clean approach would be that, e.g., the replayer thread holds the replayer and then provides the high-level information for the UI and it triggers repaints. The UI would then be very simple and stupid, i.e., doesn't have much state associated with the replayer. Maybe this idea wrongly assumed details but we should still look into checking the intention IMHO.

Reply:
I'm not sure what you mean.
The UI thread owns no state apart from the information about the viewport -> timeline mapping.
Everything else it needs for rendering (current time, cyclers, frame times) it retrieves from the replayer.

Reply:
That might be true, but my feeling is that the responsibilities are a little cluttered at the moment. My intention is to trigger another evaluation of your design. Try to optimize for minimal interfaces and minimal state. For example where should we calculate timing etc.

Reply:
Applied changes as discussed.
Please have another look.